[mlir][MemRef] Changed AssumeAlignment into a Pure ViewLikeOp

shay-kl opened the pull request:
Made AssumeAlignment a ViewLikeOp that returns a new SSA memref equal
to its memref argument and made it have Pure trait. This
gives it a defined memory effect that matches what it does in practice
and makes it behave nicely with optimizations which won't get rid of it
unless its result isn't being used.

Commit:
Added MemoryEffect to assume_alignment

Assume_alignment has no trait which specifies how it interacts with
memory, this causes an issue in OwnershipBasedBufferDeallocation,
which require all operations which operate on buffers to have explicit
traits defining how the operation interacts with memory.

From my understanding, technically the operation is pure, however to make
sure the operation doesn't get optimized away it has to have some side
effect. I defined it to have similar side effects to CF AssertOp as both
are asserts. I'm not sure if this is correct and would appreciate the
opinion of someone more experienced.

shay-kl commented:
I added a lit test to demonstrate the existing issue with the operation not currently passing OwnershipBasedBufferDeallocation, similar to the already existing lit test for cf.assert present in the same file misc-other.mlir.

MaheshRavishankar reviewed (dismissed):
Saying it has write semantics is a big pessimization. This is more  a compiler hint. If an optimization is removing it, that needs to be fixed in the optimization AFAICS. For example CSE does no CSE operation with no results (even if it has no side-effects).

matthias-springer commented:
@shay-kl Is the buffer deallocation pass happy if you use `NoMemoryEffect`?

shay-kl commented:
@MaheshRavishankar 
So you're saying that implicitly if an operation has no memory effects and no results then it makes sense for it to never be optimized away?
Would it be a better idea to explicitly define a trait for this type of operation or should we just deduct that they're an annotation and shouldn't be removed?

shay-kl commented:
@matthias-springer 
So the intent would be something like this
%res1 = memref.alloc() : memref<128xi8>
%res2 = memref.assume_alignment %res1, 64 : memref<128xi8>

and from now on we use res2 instead of res1?

joker-eph commented:
> Saying it has write semantics is a big pessimization. 

Right now without any traits, it has "unknown effect" which is most conservative situation (it has to be read and write to anything).

Commit:
Changed AssumeAlignment into a ViewLikeOp

Made AssumeAlignment a ViewLikeOp that returns a new SSA memref equal
to its memref argument and made it have NoMemoryEffect trait. This
gives it a defined memory effect that matches what it does in practice
and makes it behave nicely with optimizations which won't get rid of it
unless its result isn't being used.

shay-kl commented:
@matthias-springer 
I gave assume alignment a result and changed all its uses accordingly and it does seem cleaner this way. Thanks for the suggestion.

joker-eph commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
```diff
@@ -142,22 +142,34 @@ class AllocLikeOp<string mnemonic,
+      value. It returns a new SSA value of the same memref type, but associated
+      with the assertion that the underlying buffer is aligned to the given
+      alignment. If the buffer isn't aligned to the given alignment, the 
+      behavior is undefined.
```
Do we need undefined behavior or could we have deferred here with poison?

shay-kl replied:
I am not sure.

joker-eph replied:
That would make it Pure, which would be nice to strive for in general.

shay-kl replied:
I updated the operation to be pure and changed its description accordingly. Going through its uses in the repo it doesn't seem like I need to make changes anywhere else, but I would appreciate if someone would confirm.

joker-eph commented:
Can you update the title and description?

shay-kl commented on mlir/lib/Dialect/GPU/Transforms/EliminateBarriers.cpp:
```diff
@@ -101,9 +96,6 @@ collectEffects(Operation *op,
 
-  // Skip over ops that we know have no effects.
-  if (isKnownNoEffectsOpWithoutInterface(op))
-    return true;
```
There is no longer a need for a special case for AssumeAlignment since it now has a memory interface (NoMemoryEffect), so I removed it.

Commit:
Made AssumeAlignment Pure

Assume Alignment is now pure with its ub now deferred with poison

Commit:
Added ASM Result Naming to AssumeAlignment

Set AsmResultName for the operation to be assume_align

joker-eph commented:
LGTM but let's wait for others to double check again.

shay-kl committed:
Merge branch 'main' into memrefops/add-memory-effect-to-assume-aligned

matthias-springer commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
```diff
@@ -142,22 +142,41 @@ class AllocLikeOp<string mnemonic,
-    The `assume_alignment` operation takes a memref and an integer of alignment
-    value, and internally annotates the buffer with the given alignment. If
-    the buffer isn't aligned to the given alignment, the behavior is undefined.
+      The `assume_alignment` operation takes a memref and an integer of alignment
```
nit: drop `of`

matthias-springer commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
```diff
@@ -142,22 +142,41 @@ class AllocLikeOp<string mnemonic,
-    the buffer isn't aligned to the given alignment, the behavior is undefined.
+      The `assume_alignment` operation takes a memref and an integer of alignment
+      value. It returns a new SSA value of the same memref type, but associated
+      with the assertion that the underlying buffer is aligned to the given
```
`with the assumption`
I wouldn't call it "assertion" because it does not assert

matthias-springer commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
```diff
@@ -142,22 +142,41 @@ class AllocLikeOp<string mnemonic,
-    optimization only, and the optimization is best-effort.
+      If the buffer isn't aligned to the given alignment, this operation
+      itself does not cause undefined behavior. However, subsequent operations
+      that consume the resulting memref and rely on this asserted alignment for
```
`rely on the specified alignment`

matthias-springer commented on mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:
```diff
@@ -142,22 +142,41 @@ class AllocLikeOp<string mnemonic,
+      If the buffer isn't aligned to the given alignment, this operation
+      itself does not cause undefined behavior. However, subsequent operations
+      that consume the resulting memref and rely on this asserted alignment for
+      correctness will produce a poison value if the assertion is false.
```
I think the result of `assume_alignment` is poison if the alignment assumption is not satisfied. No need to talk about subsequent operations. See the documentation `arith.addi` for an another example that uses poison semantics.

matthias-springer reviewed (commented):
Nice!

MaheshRavishankar reviewed (commented):
THis seems good to me, but I dont fully follow why the `ViewLikeOpInterface` here?

Commit:
Modified assume_alignment description

Changed assertion to assumption and improved description of poison

shay-kl commented:
> THis seems good to me, but I dont fully follow why the `ViewLikeOpInterface` here?

I assumed that an identity operation would count as a view, and that since `assume_alignment` returns the same buffer, it could be beneficial for some generic passes. But since it doesn’t change shape or layout, I’m fine with removing the interface if that’s clearer.

shay-kl commented:
I don't have permissions, can you please land it?

AviadCo commented:
> LLVM Buildbot has detected a new failure on builder `mlir-nvidia` running on `mlir-nvidia` while building `mlir` at step 7 "test-build-check-mlir-build-only-check-mlir".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/138/builds/13299
> 
> Here is the relevant piece of the build log for the reference

Fixed failing lit test here: https://github.com/llvm/llvm-project/pull/140447
